# haptools
[![pypi version](https://img.shields.io/pypi/v/haptools)](https://pypi.org/project/haptools)
[![image](https://anaconda.org/bioconda/haptools/badges/version.svg)](https://anaconda.org/bioconda/haptools)
[![license](https://img.shields.io/pypi/l/haptools)](LICENSE)
![status](https://github.com/CAST-genomics/haptools/workflows/Tests/badge.svg)
[<img src=https://github.com/codespaces/badge.svg width=160>](https://codespaces.new/CAST-genomics/haptools)

Haptools is a collection of tools for simulating and analyzing genotypes and phenotypes while taking into account haplotype information. Haptools supports fast simulation of admixed genomes (with `simgenotype`), visualization of admixture tracks (with `karyogram`), simulating haplotype- and local ancestry-specific phenotype effects (with `transform` and `simphenotype`), and computing a variety of common file operations and statistics in a haplotype-aware manner.

Homepage: [https://haptools.readthedocs.io/](https://haptools.readthedocs.io/)

Visit our homepage for installation and usage instructions.

![haptools commands](https://github.com/CAST-genomics/haptools/assets/23412689/7de95547-3138-45c6-99d8-46bca5716c26)

## citation
There is an option to _"Cite this repository"_ on the right sidebar of [the repository homepage](https://github.com/CAST-genomics/haptools)

> Arya R Massarat, Michael Lamkin, Ciara Reeve, Amy L Williams, Matteo Dâ€™Antonio, Melissa Gymrek, Haptools: a toolkit for admixture and haplotype analysis, Bioinformatics, 2023;, btad104, https://doi.org/10.1093/bioinformatics/btad104
